### PR TITLE
add support for TMultiplexProcotol

### PR DESCRIPTION
**Pull request by lfyzjck**
Support TMultiplexProcotol in client side:

https://github.com/line/armeria/issues/240

### CODE REVIEW

**Comment by trustin** [comment]
How about this:

- Add `ClientRequestContext.fragment()` which returns the 'fragment part of the request URI,
- Instead of adding a new `ClientOption`, `THttpClientDelegate` could use the return value of `ClientRequestContext.fragment()` as the service name if the request URI had a fragment.

File: src/main/java/com/linecorp/armeria/client/thrift/THttpClientDelegate.java [hunk 1]
```diff
@@ -101,8 +103,13 @@ public ThriftReply execute(ClientRequestContext ctx, ThriftCall call) throws Exc
         }
 
         try {
+
             final TMemoryBuffer outTransport = new TMemoryBuffer(128);
-            final TProtocol tProtocol = protocolFactory.getProtocol(outTransport);
+            TProtocol tProtocol = protocolFactory.getProtocol(outTransport);
+            if (ctx.options().tMultiplex()) {
+                final String serviceName = this.path.substring(1);
+                tProtocol = new TMultiplexedProtocol(tProtocol, serviceName);
```
Comment (trustin) [inline src/main/java/com/linecorp/armeria/client/thrift/THttpClientDelegate.java]: You don't really need to use `TMultiplexedProtocol` for this purpose. You can just prepend service name to the method name:

```java
method = serviceName + ':' + method
```

**Comment by trustin** [comment]
Closing in favor of #243. Please consider working on the server side.

**Comment by lfyzjck** [comment]
TMultiplex Protocol also need put `serviceName` in url path, if use fragment instead of path, we also need to overwrite url path in `THttpClientDelegate`

**Comment by lfyzjck** [comment]
Sorry i haven't test it on the server side. I'll try to implement it on the server side.

**Comment by trustin** [comment]
@lfyzjck 

> TMultiplex Protocol also need put `serviceName` in url path, if use fragment instead of path, we also need to overwrite url path in `THttpClientDelegate`

TMultiplexedProtocol does not enforce the service name to be present in the path part of the request URI. They don't need to be same. If you want them to be same, you can just specify them twice:

```java
Clients.newClient("tbinary+http://127.0.0.1:8080/serviceName#serviceName", ...);
```

> Sorry i haven't test it on the server side. I'll try to implement it on the server side.

You can test it using your existing multiplexing Thrift servers. :-)

**Pull request closed without merging**